crates/store/src/historical/mod.rs

```diff
@@ -0,0 +1,302 @@
+        Self { block_number, rev_set }
+    }
+}
+
+#[derive(Debug)]
+struct InnerState {
+    block_number: BlockNumber,
+    latest: AccountTree,
+    overlays: VecDeque<Arc<HistoricalOverlay>>,
+}
+
+impl InnerState {
+    pub fn historical_offset(&self, desired_block_number: BlockNumber) -> HistoricalOffset {
+        let Some(past_offset) = self.block_number.checked_sub(desired_block_number.as_u32()) else {
+            return HistoricalOffset::Future;
+        };
+        let past_offset = past_offset.as_usize();
+        match past_offset {
+            0 => HistoricalOffset::Latest,
+            1..=AccountTreeWithHistory::MAX_HISTORY => {
+                HistoricalOffset::ReversionsIdx(past_offset - 1)
+            },
+            _ => HistoricalOffset::TooAncient,
+        }
+    }
+}
+
+/// wraps `AccountTree` with historical query support via reversion overlays
+#[derive(Debug, Clone)]
+pub struct AccountTreeWithHistory {
+    inner: Arc<RwLock<InnerState>>,
+}
+
+impl AccountTreeWithHistory {
+    pub const MAX_HISTORY: usize = 33;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// creates new historical tree
+    pub fn new(account_tree: AccountTree, block_number: BlockNumber) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(InnerState {
+                block_number,
+                latest: account_tree,
+                overlays: VecDeque::new(),
+            })),
+        }
+    }
+
+    fn cleanup(overlays: &mut VecDeque<Arc<HistoricalOverlay>>) {
+        while overlays.len() > Self::MAX_HISTORY {
+            overlays.pop_back();
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// returns latest block number
+    pub fn block_number(&self) -> BlockNumber {
+        self.inner.read().unwrap().block_number
+    }
+
+    /// returns latest root
+    pub fn root(&self) -> Word {
+        self.inner.read().unwrap().latest.root()
+    }
+
+    /// returns root at given historical block
+    pub fn root_at(&self, block_number: BlockNumber) -> Option<Word> {
+        let guard = self.inner.read().unwrap();
+
+        match guard.historical_offset(block_number) {
+            HistoricalOffset::Latest => Some(guard.latest.root()),
+            HistoricalOffset::ReversionsIdx(idx) => {
+                let overlay_at = guard.overlays.get(idx)?;
+                assert_eq!(overlay_at.block_number, block_number);
+                Some(overlay_at.rev_set.as_mutation_set().root())
+            },
+            HistoricalOffset::Future | HistoricalOffset::TooAncient => None,
+        }
+    }
+
+    /// returns account count
+    pub fn num_accounts(&self) -> usize {
+        self.inner.read().unwrap().latest.num_accounts()
+    }
+
+    /// returns history depth
```

Comment:
nit: capital letter + full stop for all doc comments. (and comments in general?)